<span style="color:#1871d6; font-size:24px; font-weight:700"> Conditional Probability</span>

In this lab notebook we will be looking into independent and dependent events, 
concepts of conditional probablity, permutations, and combinations. 

Independent and Dependent events follow from their names. 
Independent events are those whose occurence is wholly independent of another event, 
or at least the events we are calculating probabilities for. 
Dependent events exist in a space where the likliehood of one event depends on or is influenced by the occurrence of the other. 
We express this below mathematically. 

Conditional probability is the probability of seeing some event knowing that some other event has actually occurred. For example, weather forecasting is based on conditional probabilities. 
When the forecast says that there is a 30% chance of rain, 
that probability is based on all the information that the meteorologists know up until that point.

Let's look into what independent and dependent events are. 
**Reference: ** [Elementary Probability and the prob Package](https://cran.r-project.org/web/packages/prob/vignettes/prob.pdf)

<span style="color:#1871d6; font-size:20px; font-weight:500">  Independent and Dependent Events</span>


**Independent: ** The probabilities of events A and B are said to be statistically independent if the fact that one event has occurred does not affect the probability that the other event will occur.  The multiplication rule for independent events is:

                                          P(A ∩ B) = P(A) * P(B)     
            
           
       IF P(A) * P(B) = P(A and B) THEN A and B are independent events, otherwise, they are dependent events.

If the occurrence of Event A changes the probability of Event B, then Events A and B are **dependent.**  

A **conditional probability** is one that denotes the probability that Event A occurs, given that Event B has occurred.  Here is how we represent a conditional probability:
    
                                                     P(A ∩ B)    
                                            P(A|B) = -------- 
                                                       P(B)         
                                                     

In a conditional probability, **if A and B are independent** (which then means it's not truly a conditional probability), the numerator of the fraction will factors out so that P(B) then cancels out:
 
                       P(A ∩ B)                                        P(A) * P(B)
              P(A|B) = --------    if A & B are independent     =>  =  ----------- =   P(A)
                         P(B)                                              P(B)
                                                     

Let's consider the example of tossing ten coins to illustrate the nature of independent events. 
So, what is the probability of observing at least one Head? 

Imagine that we are tossing the coins in such a way that they do not interfere with each other; i.e. they are independent events. 

The only way there will not be at least one Head is if all tosses are Tails. 
Therefore,
         
         P(at least one H) = 1 − P(all T),


                             1 - (1/2)^10

In [2]:
#Let's try to do the same thing in R

library(prob)

Space <- tosscoin(10, makespace = TRUE)

# The isrep function in the prob package, will test each row of Space to see whether the value T appears 10 times 
# and returns true or false for each row it checks. The subset function is logical, so makes a subset with the 
# rows which are true.

head(isrep(Space, vals = "T", nrep = 10))

subset(Space, isrep(Space, vals = "T", nrep = 10))

A <- subset(Space, isrep(Space, vals = "T", nrep = 10))

1 - Prob(A)

[1] FALSE FALSE FALSE FALSE FALSE FALSE

toss1 toss2 toss3 toss4 toss5 toss6 toss7 toss8 toss9 toss10 probs       
1024 T     T     T     T     T     T     T     T     T     T      0.0009765625

[1] 0.9990234

<span style="color:#1871d6; font-size:16px; font-weight:700"> Repeated Experiments with Independent Events </span>

Experiments are repeated when we want to discern the probability of two events occuring more reliably. 
Often, a single experiment does not yield sufficient data. 
Therefore, it is common to repeat a certain experiment multiple times under identical 
conditions and in an independent manner. 
Experiments like tossing a coin repeatedly, rolling a die or dice, etc. are repeated experiments.

The `iidspace` function in the "prob" library in R (note "*`library(prob)`* in the code above) implements repeated experiments. 
It takes three arguments: 
`x`, which is a vector of outcomes, 
`ntrials`, which is an integer telling how many times to repeat the experiment, and 
`probs` to specify the probabilities of the outcomes of x in a single trial.

In [3]:
iidspace(c("H","T"), ntrials = 3, probs = c(0.5, 0.5))

X1 X2 X3 probs
1 H  H  H  0.125
2 T  H  H  0.125
3 H  T  H  0.125
4 T  T  H  0.125
5 H  H  T  0.125
6 T  H  T  0.125
7 H  T  T  0.125
8 T  T  T  0.125

In [4]:
iidspace(c("H","T"), ntrials = 7, probs = c(0.5, 0.5))

X1 X2 X3 X4 X5 X6 X7 probs    
1   H  H  H  H  H  H  H  0.0078125
2   T  H  H  H  H  H  H  0.0078125
3   H  T  H  H  H  H  H  0.0078125
4   T  T  H  H  H  H  H  0.0078125
5   H  H  T  H  H  H  H  0.0078125
6   T  H  T  H  H  H  H  0.0078125
7   H  T  T  H  H  H  H  0.0078125
8   T  T  T  H  H  H  H  0.0078125
9   H  H  H  T  H  H  H  0.0078125
10  T  H  H  T  H  H  H  0.0078125
11  H  T  H  T  H  H  H  0.0078125
12  T  T  H  T  H  H  H  0.0078125
13  H  H  T  T  H  H  H  0.0078125
14  T  H  T  T  H  H  H  0.0078125
15  H  T  T  T  H  H  H  0.0078125
16  T  T  T  T  H  H  H  0.0078125
17  H  H  H  H  T  H  H  0.0078125
18  T  H  H  H  T  H  H  0.0078125
19  H  T  H  H  T  H  H  0.0078125
20  T  T  H  H  T  H  H  0.0078125
21  H  H  T  H  T  H  H  0.0078125
22  T  H  T  H  T  H  H  0.0078125
23  H  T  T  H  T  H  H  0.0078125
24  T  T  T  H  T  H  H  0.0078125
25  H  H  H  T  T  H  H  0.0078125
26  T  H  H  T  T  H  H  0.0078125
27  H  T  H  T  T  H  H  0.0078125
28  T  T  H  T  T  H  H  0.0078125
29  H  H  T  T  T  H  H  0.0078125
30  T  H  T  T  T  H  H  0.0078125
⋮   ⋮  ⋮  ⋮  ⋮  ⋮  ⋮  ⋮  ⋮        
99  H  T  H  H  H  T  T  0.0078125
100 T  T  H  H  H  T  T  0.0078125
101 H  H  T  H  H  T  T  0.0078125
102 T  H  T  H  H  T  T  0.0078125
103 H  T  T  H  H  T  T  0.0078125
104 T  T  T  H  H  T  T  0.0078125
105 H  H  H  T  H  T  T  0.0078125
106 T  H  H  T  H  T  T  0.0078125
107 H  T  H  T  H  T  T  0.0078125
108 T  T  H  T  H  T  T  0.0078125
109 H  H  T  T  H  T  T  0.0078125
110 T  H  T  T  H  T  T  0.0078125
111 H  T  T  T  H  T  T  0.0078125
112 T  T  T  T  H  T  T  0.0078125
113 H  H  H  H  T  T  T  0.0078125
114 T  H  H  H  T  T  T  0.0078125
115 H  T  H  H  T  T  T  0.0078125
116 T  T  H  H  T  T  T  0.0078125
117 H  H  T  H  T  T  T  0.0078125
118 T  H  T  H  T  T  T  0.0078125
119 H  T  T  H  T  T  T  0.0078125
120 T  T  T  H  T  T  T  0.0078125
121 H  H  H  T  T  T  T  0.0078125
122 T  H  H  T  T  T  T  0.0078125
123 H  T  H  T  T  T  T  0.0078125
124 T  T  H  T  T  T  T  0.0078125
125 H  H  T  T  T  T  T  0.0078125
126 T  H  T  T  T  T  T  0.0078125
127 H  T  T  T  T  T  T  0.0078125
128 T  T  T  T  T  T  T  0.0078125

<span style="color:#1871d6; font-size:20px; font-weight:700"> Dependent Events / Conditional Probability

Conditional probabilities only exist for dependent events. 
Consider why: 
If events are independent, then their occurence probabilities cannot be expressed together in an equation. 
If events are *dependent*, then their probabilities can be expressed as one depending on the other or in an expression of mutual dependence. 

Consider an example of drawing cards from a full deck of 52 standard playing cards as an example of *dependent events* and *conditional probability*. 
Select two cards from the deck, in succession. 

                   Let A = {first card drawn is an Ace} and B = {second card drawn is an Ace}. 

Since there are four Aces in the deck, it is natural to assign P(A) = 4/52. 
Let's unpack how this probability changes after the first card is drawn,
because after the first card is drawn there are only 51 cards remaining. 

Suppose we looked at the first card. 
What is the probability of B now? 
The answer depends on the value of the first card. 
If the first card is an Ace, then the probability that the second also is an Ace should be 3/51, 
but if the first card is not an Ace, then the probability that the second is an Ace should be 4/51. 

Mathematically, for the situation in which Event B is a drawn Ace, after Event A is a drawn Ace:
      
                                       P(B|A) = 3/51, P(A) = 4/52
    
The probability of B being an Ace, if A was an Ace is 3/51. 
The probability of A being an Ace is 4/52, because no cards have been drawn yet. 
    
Definition: The conditional probability of B given A (i.e., the probability of B given that A occurs), denoted P(B|A), is defined by:
    
    
                                                  P(A ∩ B)
                                        P(B|A) =  --------
                                                    P(A)
                
We can't factor the numerator and cancel the P(A)'s here, though, because the events are not independent.

P(A ∩ B) means _the probability that A & B intersect_ -- or the probability that both A and B occur.

In this case, they both occur if both cards are Aces.  The events are _dependent_ because the occurrence of Event A changes the probability of the occurrence of Event B.  P(A) represents the probability that Event A is a drawn Ace.
    
                       P(A ∩ B)                                             3      4
             P(B|A) =  --------       =>       P(A ∩ B) = P(B|A) * P(A) => --- *  --- => 0.0045
                         P(A)                                               51     52

In [5]:
(3/51)*(4/52)

[1] 0.004524887

**Example:** Let's work out an example. 
Toss a six-sided die _twice_. 
The sample space consists of all ordered pairs (i, j)

of the numbers 1, 2, . . . , 6, that is, S = {(1, 1), (1, 2), . . . ,(6, 6)}. 

Essentially, "i" is one die and "j" is another die. 

Let A = {outcomes match} and B = {sum of outcomes at least 8}.

In [6]:
# The first thing to do is set up the probability space with the 
# rolldie function inside the prob library. 
# So, "S" is the probability space
# "rolldie" is the function
# 2 is the number of die
# makespace is "make the space? Yes or no?"
S <- rolldie(2, makespace = TRUE)
# S contains all the all 36 possible outcomes {(1,1),(1,2)....(6,6)} with 
# each outcome having an identical probability of 0.02777778
head(S)

X1 X2 probs     
1 1  1  0.02777778
2 2  1  0.02777778
3 3  1  0.02777778
4 4  1  0.02777778
5 5  1  0.02777778
6 6  1  0.02777778

In [7]:
# Subsetting sample space S, for outcomes matching event A(outcomes match). 
# This results in a set where both die are the same (i & j are the same)
A <- subset(S, X1 == X2)
A

X1 X2 probs     
1  1  1  0.02777778
8  2  2  0.02777778
15 3  3  0.02777778
22 4  4  0.02777778
29 5  5  0.02777778
36 6  6  0.02777778

In [8]:
# Subsetting sample space S, 
# for outcomes matching event B(sum of outcomes at least 8). 
# The die total must be 8 or more
B <- subset(S, X1 + X2 >= 8)
B

X1 X2 probs     
12 6  2  0.02777778
17 5  3  0.02777778
18 6  3  0.02777778
22 4  4  0.02777778
23 5  4  0.02777778
24 6  4  0.02777778
27 3  5  0.02777778
28 4  5  0.02777778
29 5  5  0.02777778
30 6  5  0.02777778
32 2  6  0.02777778
33 3  6  0.02777778
34 4  6  0.02777778
35 5  6  0.02777778
36 6  6  0.02777778

In [9]:
# When calculating conditional probability, we should use the 
# "given" argument of the prob function as shown below:
# A is the event of getting same outcome {{1,1},{2,2}...{6,6}}
# B is the event of getting the outcomes with sum of >=8 {{2,6},{3,6},{4,6},{5,6},{6,6},{3,5}...{6,6}}
paste('P(A/B): ',Prob(A, given = B))

paste('P(B/A): ',Prob(B, given = A))

# Instead of defining events A and B you can directly do conditional probability, if we reference the original 
# probability space S as the first argument of the prob calculation as shown below:
paste('P(A/B): ', Prob(S, X1 == X2, given = (X1 + X2 >= 8)))

paste('P(B/A): ', Prob(S, X1 + X2 >= 8, given = (X1 == X2)))

[1] "P(A/B):  0.2"

[1] "P(B/A):  0.5"

[1] "P(A/B):  0.2"

[1] "P(B/A):  0.5"

The above examples shown are simple applications of conditional probability on a die. 
`prob` package can be extended to multivariate datasets where events can be defined 
as columns and supplied as arguments, like in the previous examples.

<span style="color:#1871d6; font-size:20px; font-weight:700"> Permutations and Combinations

The main difference between combinations and permutations is that a combination does not take into account the order, whereas a permutation does.

Consider a simple example from [mathisfun](http://www.mathsisfun.com/combinatorics/combinations-permutations.html). 
When we say "My fruit salad is a **combination** of apples, grapes and bananas", we are not bothered about what order the fruits are in. No matter in which order you mention the fruits, it's the same fruit salad.

But when we say "You need the combination 123 to open the safe", 
we care about the order of numbers. 
No other combination will work to open the safe. 
It has to be exactly 1-2-3. 
This is a **permutation**.

  * When the order doesn't matter, it is a Combination.
	
  * When the order does matter, it is a Permutation.
    

There are many ways you can create permutations and combinations in R. 
We will be using combinat package for this. 

<span style="color:#e57f0b; font-size:14px; font-weight:700">combn():</span>

`combn()` is used to generate combinations. Its usage is illustrated below. 

`Usage`

    combn(x, m, fun=NULL, simplify=TRUE, ...)


`Arguments`

    x         vector source for combinations i'e the vector of elements used to generate the combinations 
    m         number of elements in combination. If you specify 2 as input, combinations of size two are generated.
    fun       function to be applied to each combination (may be null). It can be any function like sum(), mean() etc.
    simplify  logical, if FALSE, returns a list, otherwise returns vector or array. 
    ...       args to fun

It generates all combinations of the elements of x taken m at a time. 
In code snippet below, we have given an input of 4 to x and 2 to m. 
So, the function has to return combinations of size 2 using the numbers {1,2,3,4}, like {{1,2},{1,3}....}. 

If argument FUN is not NULL, the code applies a function given by the argument to each point. 
We will supply sum() as the function. 
If `simplify` is FALSE, it returns a list; otherwise, it returns an array, typically a matrix. 
"..." are passed unchanged to the FUN function, if specified.

In [10]:
library(combinat)
#Generate different possible combinations of size 2 using numbers {1,2,3,4}
combn(4, 2)

print("sum of elements of each combination ")
#Generate different possible combinations of size 2 using numbers {1,2,3,4} and return their sums.
combn(4, 2,sum)

[,1] [,2] [,3] [,4] [,5] [,6]
[1,] 1    1    1    2    2    3   
[2,] 2    3    4    3    4    4

[1] "sum of elements of each combination "


[1] 3 4 5 5 6 7

**permn(): ** `permn()` is used to generate permutations. 

`Usage`

    permn(x, fun=NULL, ...)
        
        
`Arguments`

    x    vector source for permutations i'e the vector of elements used to generate the permutations 
    fun  if non.null, applied at each perm

Generates all permutations of the elements of x. 
In the example below we have given 3 as our input in order to generate permutations of size 3, like {{1,2,3},{1,3,2},{2,1,3}...} etc. 
If argument "fun" is not null, it applies a function given by the argument to each point. 

In [11]:
#Generate different possible permutations using numbers (1,2,3)
permn(3)

#Generate different possible permutations using numbers (1,2,3) and return standard deviation of permutations.
permn(3,sd)

[[1]]
[1] 1 2 3

[[2]]
[1] 1 3 2

[[3]]
[1] 3 1 2

[[4]]
[1] 3 2 1

[[5]]
[1] 2 3 1

[[6]]
[1] 2 1 3

[[1]]
[1] 1

[[2]]
[1] 1

[[3]]
[1] 1

[[4]]
[1] 1

[[5]]
[1] 1

[[6]]
[1] 1

In [12]:
# You can find the number of permutations generated using length function. 
length(permn(3))

[1] 6

### Extensions of probability to multivariate data

We have seen how conditional probability has been applied to simple dice events. 
Let's continue our discussion to multivariate data. 
We will work with the motor vehicle thefts dataset. 
The data is a combination of both factor and continuous variables. 
The table() command is used extensively when dealing with conditional probability.

Load the dataset into a dataframe called `vehicle_thefts`. 
Dataset is located in '/dsa/data/all_datasets/motor_vehicle_thefts/' directory. 

**NOTE:** This is a variation of the mvt.csv file that is used in other courses.

In [13]:
vehicle_thefts <- read.csv("/dsa/data/all_datasets/motor_vehicle_thefts/mvt.csv", header = TRUE)

head(vehicle_thefts)

ID      Date             LocationDescription           Arrest Domestic Beat
1 8951354 12/31/2012 23:15 STREET                        FALSE  FALSE     623
2 8951141 12/31/2012 22:00 STREET                        FALSE  FALSE    1213
3 8952745 12/31/2012 22:00 RESIDENTIAL YARD (FRONT/BACK) FALSE  FALSE    1622
4 8952223 12/31/2012 22:00 STREET                        FALSE  FALSE     724
5 8951608 12/31/2012 21:30 STREET                        FALSE  FALSE     211
6 8950793 12/31/2012 20:30 STREET                         TRUE  FALSE    2521
  District CommunityArea Year Latitude Longitude
1  6       69            2012 41.75628 -87.62164
2 12       24            2012 41.89879 -87.66130
3 16       11            2012 41.96919 -87.76767
4  7       67            2012 41.76933 -87.65773
5  2       35            2012 41.83757 -87.62176
6 25       19            2012 41.92856 -87.75400

In [14]:
# Extract month, weekday, hour etc. values from a date variable. Convert the format of date variable into a 
# standard format so that day, month, year etc. values can be extracted from a date.

# There are two internal implementations of date/time: POSIXct, which stores seconds since UNIX 
# epoch (+ some other data), and POSIXlt, which stores a list of day, month, year, hour, minute, second, etc.

# strptime is a function to directly convert character vectors (of a variety of formats) to POSIXlt format.
# In this dataset, the date variable was originally stored as a character vector.

DateConvert = strptime(vehicle_thefts$Date, "%m/%d/%Y")

# Extract the month and the day of the week and add these variables to the data frame vehicle_thefts.
# months() and weekdays() functions help you extract the values from a "POSIXlt" object

vehicle_thefts$Month <- months(DateConvert)

vehicle_thefts$Weekday <- weekdays(DateConvert)

**Reference:** [strptime()](http://rfunction.com/archives/1912)                

In [15]:
head(vehicle_thefts)

ID      Date             LocationDescription           Arrest Domestic Beat
1 8951354 12/31/2012 23:15 STREET                        FALSE  FALSE     623
2 8951141 12/31/2012 22:00 STREET                        FALSE  FALSE    1213
3 8952745 12/31/2012 22:00 RESIDENTIAL YARD (FRONT/BACK) FALSE  FALSE    1622
4 8952223 12/31/2012 22:00 STREET                        FALSE  FALSE     724
5 8951608 12/31/2012 21:30 STREET                        FALSE  FALSE     211
6 8950793 12/31/2012 20:30 STREET                         TRUE  FALSE    2521
  District CommunityArea Year Latitude Longitude Month    Weekday
1  6       69            2012 41.75628 -87.62164 December Monday 
2 12       24            2012 41.89879 -87.66130 December Monday 
3 16       11            2012 41.96919 -87.76767 December Monday 
4  7       67            2012 41.76933 -87.65773 December Monday 
5  2       35            2012 41.83757 -87.62176 December Monday 
6 25       19            2012 41.92856 -87.75400 December Monday

In [16]:
# What is the probability that an arrest has occurred for domestic motor vehicle theft?

# We have to find the distribution of thefts based on whether they are domestic or not. We will use a 2-way 
# table to generate these frequencies and to use these frequencies to determine probabilities.

# with() is used to avoid refering to the dataframe every time we refer to one of its variables in the table command. 
with(vehicle_thefts, table(Arrest, Domestic))

       Domestic
Arrest   FALSE   TRUE
  FALSE 175755    350
  TRUE   15471     65

In [17]:
# We are trying to find out the probability of arrest happening given the theft is a Domestic type. 
# Mathematically, this is represented as P(Arrest|Domestic) = P(Arrest & Domestic)/P(Domestic) = 65 / 415

# P(Arrest & Domestic) = 65. Look at above table for the instances where arrest is TRUE and domestic is TRUE.
# P(Domestic) = 415. Look for instances where Domestic is true in above table.

65 / (350 + 65)

[1] 0.1566265

**Reference: ** [with()](http://www.statmethods.net/stats/withby.html). 

You can use `attach()` function as an alternative to `with()`. 
`attach()` is used to make objects within dataframes accessible in R with fewer keystrokes. 
Once you attach the dataframe you can refer to its variable without referring to the dataframe.

**Reference: ** [attach()](https://www.r-bloggers.com/to-attach-or-not-attach-that-is-the-question/)

In [18]:
# what is the probability that an arrest has been made for motor thefts in the year 2001?
with(vehicle_thefts,table(Arrest, Year))

       Year
Arrest   2001  2002  2003  2004  2005  2006  2007  2008  2009  2010  2011  2012
  FALSE 18517 16638 14859 15169 14956 14796 13068 13425 11327 14796 15012 13542
  TRUE   2152  2115  1798  1693  1528  1302  1212  1020   840   701   625   550

In [19]:
# P(Arrest|Year == 2001) = P(Arrest & Year == 2001)/P(Year == 2001)

# P(Arrest being made & Year == 2001) = 2152
# P(Year == 2001) = 2152 + 18517

2152 / (2152 + 18517)

[1] 0.1041173

In [20]:
# There are different locations where the cars are being stolen from. Subset the data using top 5 locations in 
# the order of maximum number of thefts, excluding the "Other" category.  Select the bottom 5 of the following 
# options.

sort(table(vehicle_thefts$LocationDescription))


    AIRPORT BUILDING NON-TERMINAL - SECURE AREA 
                                              1 
                 AIRPORT EXTERIOR - SECURE AREA 
                                              1 
                                ANIMAL HOSPITAL 
                                              1 
                                APPLIANCE STORE 
                                              1 
                                      CTA TRAIN 
                                              1 
                        JAIL / LOCK-UP FACILITY 
                                              1 
                                      NEWSSTAND 
                                              1 
                                         BRIDGE 
                                              2 
              COLLEGE/UNIVERSITY RESIDENCE HALL 
                                              2 
                              CURRENCY EXCHANGE 
                                              2 
                   

In [21]:
# Create a subset of data, including observations for which the theft happened in one of the top five locations.  
# Call this new data set "Top5".

Top5 <- subset(vehicle_thefts, vehicle_thefts$LocationDescription=="STREET" | 
                               vehicle_thefts$LocationDescription=="PARKING LOT/GARAGE(NON.RESID.)" | 
                               vehicle_thefts$LocationDescription=="ALLEY" | 
                               vehicle_thefts$LocationDescription=="DRIVEWAY - RESIDENTIAL" | 
                               vehicle_thefts$LocationDescription=="GAS STATION")


str(Top5)

'data.frame':	177510 obs. of  13 variables:
 $ ID                 : int  8951354 8951141 8952223 8951608 8950793 8950760 8951611 8951802 8950706 8951585 ...
 $ Date               : Factor w/ 131679 levels "1/1/2001 0:01",..: 42824 42823 42823 42822 42821 42820 42819 42817 42816 42816 ...
 $ LocationDescription: Factor w/ 78 levels "ABANDONED BUILDING",..: 72 72 72 72 72 72 72 72 72 72 ...
 $ Arrest             : logi  FALSE FALSE FALSE FALSE TRUE FALSE ...
 $ Domestic           : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...
 $ Beat               : int  623 1213 724 211 2521 423 231 1021 1215 1011 ...
 $ District           : int  6 12 7 2 25 4 2 10 12 10 ...
 $ CommunityArea      : int  69 24 67 35 19 48 40 29 24 29 ...
 $ Year               : int  2012 2012 2012 2012 2012 2012 2012 2012 2012 2012 ...
 $ Latitude           : num  41.8 41.9 41.8 41.8 41.9 ...
 $ Longitude          : num  -87.6 -87.7 -87.7 -87.6 -87.8 ...
 $ Month              : chr  "December" "December" "December" "Dec

Take a look at the number of levels of **LocationDescription**. 
Ideally, the new dataframe `Top5` should contain only five locations: 
STREET, PARKING LOT/GARAGE(NON.RESID.), ALLEY, DRIVEWAY - RESIDENTIAL and GAS STATION. 
However, str() says **LocationDescription** has 78 levels.

So, what's going on here?

In [22]:
# R will remember the other categories of the LocationDescription variable from the original dataset 
# 'vehicle_thefts'. Therefore, update the LocationDescription of Top5 dataframe according to new data. If 
# you forget to update the LocationDescription, the Top5$LocationDescription will contain all 78 levels 
# that you find in vehicle_thefts$LocationDescription.

Top5$LocationDescription = factor(Top5$LocationDescription)

str(Top5)

'data.frame':	177510 obs. of  13 variables:
 $ ID                 : int  8951354 8951141 8952223 8951608 8950793 8950760 8951611 8951802 8950706 8951585 ...
 $ Date               : Factor w/ 131679 levels "1/1/2001 0:01",..: 42824 42823 42823 42822 42821 42820 42819 42817 42816 42816 ...
 $ LocationDescription: Factor w/ 5 levels "ALLEY","DRIVEWAY - RESIDENTIAL",..: 5 5 5 5 5 5 5 5 5 5 ...
 $ Arrest             : logi  FALSE FALSE FALSE FALSE TRUE FALSE ...
 $ Domestic           : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...
 $ Beat               : int  623 1213 724 211 2521 423 231 1021 1215 1011 ...
 $ District           : int  6 12 7 2 25 4 2 10 12 10 ...
 $ CommunityArea      : int  69 24 67 35 19 48 40 29 24 29 ...
 $ Year               : int  2012 2012 2012 2012 2012 2012 2012 2012 2012 2012 ...
 $ Latitude           : num  41.8 41.9 41.8 41.8 41.9 ...
 $ Longitude          : num  -87.6 -87.7 -87.7 -87.6 -87.8 ...
 $ Month              : chr  "December" "December" "December" "De

In [23]:
# What is the probability that an arrest is made and the place is street?

with(Top5,table(LocationDescription, Arrest))

                                Arrest
LocationDescription               FALSE   TRUE
  ALLEY                            2059    249
  DRIVEWAY - RESIDENTIAL           1543    132
  GAS STATION                      1672    439
  PARKING LOT/GARAGE(NON.RESID.)  13249   1603
  STREET                         144969  11595

In [24]:
# P(arrest|LocationDescription == "street") = P(Arrest & street)/P(street)

# P(Arrest being made & location is 'street') = 11595

# P(street) = 11595 + 144969 
 
(11595) / (144969 + 11595)

[1] 0.07405917

In [25]:
# What is the probability that an arrest did not happen and the weekday is Monday?

with(Top5, table(Weekday, Arrest))

           Arrest
Weekday     FALSE  TRUE
  Friday    24929  2149
  Monday    23334  1954
  Saturday  23213  2042
  Sunday    22421  2135
  Thursday  23374  1864
  Tuesday   22917  1880
  Wednesday 23304  1994

In [26]:
# P(!Arrest|Weekday == "monday") = P(!Arrest & Weekday)/P(Weekday)

# P(Arrest didn't happen & Weekday is 'Monday') = 23334
# P(Weekday  is 'monday') = 23334 + 1954

23334 / (23334 + 1954)

[1] 0.9227301

# This concludes the lab on Conditional Probability